File: src/features/typing/data/java/longest_substring_without_repeating_characters.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> last = new HashMap<>();
        int start = 0;
        int best = 0;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            Integer prev = last.get(c);
            if (prev != null && prev >= start) {
                start = prev + 1;
            }
            last.put(c, i);
            best = Math.max(best, i - start + 1);
        }
        return best;
    }
}

File: src/features/typing/data/dart/longest_substring_without_repeating_characters.dart
class Solution {
  int lengthOfLongestSubstring(String s) {
    final last = <String, int>{};
    var start = 0;
    var best = 0;
    for (var i = 0; i < s.length; i++) {
      final c = s[i];
      final prev = last[c];
      if (prev != null && prev >= start) {
        start = prev + 1;
      }
      last[c] = i;
      if (i - start + 1 > best) {
        best = i - start + 1;
      }
    }
    return best;
  }
}
